homeassistant/components/opnsense/config_flow.py

```diff
@@ -0,0 +1,138 @@
+"""Config flow for OPNsense."""
+
+import logging
+from typing import Any
+
+from pyopnsense import diagnostics
+from pyopnsense.exceptions import APIException
+from requests.exceptions import ConnectionError as requestsConnectionError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.config_entries import ConfigFlowResult
+from homeassistant.const import (
+    CONF_API_KEY,
+    CONF_HOST,
+    CONF_PORT,
+    CONF_SSL,
+    CONF_URL,
+    CONF_VERIFY_SSL,
+)
+
+from .const import CONF_API_SECRET, CONF_TRACKER_INTERFACE, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class OPNsenseConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """OPNsense config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+    data: Any | None = None
+
+    async def _show_setup_form(
+        self,
+        user_input: (dict[Any, Any] | None) = None,
+        errors: (dict[Any, Any] | None) = None,
+    ) -> ConfigFlowResult:
+        """Show the setup form to the user."""
+        if user_input is None:
+            user_input = {}
+
+        description_placeholders = {}
+        description_placeholders["doc_url"] = (
+            "https://www.home-assistant.io/integrations/opnsense/"
+        )
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOST, default=user_input.get(CONF_HOST, "")): str,
+                    vol.Required(
+                        CONF_PORT, default=user_input.get(CONF_PORT, 80)
+                    ): vol.Coerce(int),
+                    vol.Required(
+                        CONF_API_KEY, default=user_input.get(CONF_API_KEY, "")
+                    ): str,
+                    vol.Required(
+                        CONF_API_SECRET, default=user_input.get(CONF_API_SECRET, "")
+                    ): str,
+                    vol.Required(
+                        CONF_SSL, default=user_input.get(CONF_SSL, False)
+                    ): bool,
+                    vol.Required(
+                        CONF_VERIFY_SSL, default=user_input.get(CONF_VERIFY_SSL, False)
+                    ): bool,
+                    vol.Optional(CONF_TRACKER_INTERFACE): str,
+                }
+            ),
+            errors=errors or {},
+            description_placeholders=description_placeholders,
+        )
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle user step."""
+        errors = {}
+
+        if user_input is None:
+            return await self._show_setup_form(user_input, None)
+
+        await self.async_set_unique_id(user_input[CONF_API_KEY])
```

Comment:
Is it common to use an API KEY as a unique ID? This seems to be a recipe to expose that secret where we don't want to.

Reply:
If you can generate a new api key, the api key is a bad unique identifier, if there is nothing better, you should use `self._abort_entries_match`

Reply:
Could we use a combination of the name of this integration at the `host` input from the user? It would then be something like `"opnsense_192.168.1.1"` or `"opnsense_router.myhome.net:8443"`

Reply:
No sounds like we shouldn't use that either based on https://developers.home-assistant.io/docs/entity_registry_index/